server/src/main/java/org/elasticsearch/common/cache/Cache.java

```diff
@@ -650,6 +650,36 @@ public void remove() {
         };
     }
 
+    /**
+     * An LRU sequencing of the entries in the cache. This sequence is not protected from mutations
+     * to the cache (except for {@link Iterator#remove()}. The result of iteration under any other mutation is
+     * undefined.
```

Comment:
I think this is true. In fact, it looks like the lru-chain would be unsafely published with no happens-before to the reader.

I think the iteration of the entries done in this PR is thus unsafe, at least it might skip some entries.

I think the same applies to iterating over `keys()`, which we do a few places.

Together with other comments in this PR, this makes me think that perhaps it was easier to let `CacheFile`'s that need fsync register with the `CacheService` (or a registry in between)?